### PR TITLE
Add code generator for plain grpc server

### PR DESCRIPTION
**Pull request by r3mariano**
The goal with this change is to allow using wire classes with the plain grpc Netty server. This is the code generator counterpart to the PoC in Misk: https://github.com/cashapp/misk/pull/1706

Things to do:

- [x] complete generated file
- [x] test generator components
- [x] hook generator to KotlinTarget?

### CODE REVIEW

File: wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/AddBlockingStub.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      .build()
+  )
+  .addType(
+    TypeSpec.classBuilder("${service.name}BlockingStub")
+      .addAbstractStubConstructor(service)
+      .addBlockingStubRpcCalls(service)
+      .build()
+  )
+
+private fun TypeSpec.Builder.addBlockingStubRpcCalls(service: Service): TypeSpec.Builder {
```
Comment (mightyguava) [inline wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/AddBlockingStub.kt:30]: These seem like strange uses of extension functions. Extension functions are meant for adding reusable methods to existing classes, whereas it seems here you are using it as a way to enable method chaining. It makes the implementation a bit hard to follow. I think you should be creating methods on your KotlinGrpcGenerator class that take `TypeSpec.Builder` as an argument instead, or separate "Generator" classes for each component if one is too big

File: wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSinkAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.squareup.wire.kotlin.grpcserver
```
Comment (mightyguava) [inline wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSinkAdapter.kt:1]: These library classes should probably be in a standalone module.

File: wire-library/wire-grpc-java-server/src/test/java/com/squareup/wire/kotlin/grpcserver/AddImplBaseKtTest.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package com.squareup.wire.kotlin.grpcserver
```
Comment (mightyguava) [inline wire-library/wire-grpc-java-server/src/test/java/com/squareup/wire/kotlin/grpcserver/AddImplBaseKtTest.kt:1]: can we do golden tests instead of just asserting that a few lines match? Is there precedence in this repo?

**Review by mightyguava — COMMENTED** [review]
Great work! Let's figure out with the maintainers here the next steps. I imagine this will take some iteration

File: wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/AddBlockingStub.kt (hunk 1 above)
Comment (r3mariano) [reply wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/AddBlockingStub.kt:30]: That's a fair point about using extension functions here. 

Each of the parts of the generated code are a bit big and I'd like to test them separately. I don't think we should use a class to house a single method though, kotlin allows for top-level functions after all.

File: wire-library/wire-grpc-java-server/src/test/java/com/squareup/wire/kotlin/grpcserver/AddImplBaseKtTest.kt (hunk 3 above)
Comment (r3mariano) [reply wire-library/wire-grpc-java-server/src/test/java/com/squareup/wire/kotlin/grpcserver/AddImplBaseKtTest.kt:1]: The precedence in Kotlin and Java generators is to check that a few lines match. I agree though, it would be much better to do golden files. I'll go add them.

File: wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/AddBlockingStub.kt (hunk 1 above)
Comment (mightyguava) [reply wire-library/wire-grpc-java-server/src/main/java/com/squareup/wire/kotlin/grpcserver/AddBlockingStub.kt:30]: You could make several classes, each responsible for one of the components that get generated? I'd just follow the patterns in https://github.com/square/wire/blob/master/wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package com.squareup.wire.kotlin.grpcserver
```
Comment (oldergod) [inline wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt:1]: You'll wanna add copyright to most new files

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/LegacyAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    service.rpcs.forEach { rpc ->
+      val codeBlock = when {
+        rpc.requestStreaming && rpc.responseStreaming -> CodeBlock.of(
+          """
+          val requestStream = %T()
+          streamExecutor.submit {
+            ${rpc.name}().${rpc.name}(requestStream, %T(response))
+          }
+          return requestStream
+        """.trimIndent(),
```
Comment (oldergod) [inline wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/LegacyAdapter.kt:100]: indention is weird

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      )
+
+  private fun TypeSpec.Builder.addBlockingStubRpcCalls(service: Service): TypeSpec.Builder {
+    service.rpcs
+      .filter { !it.responseStreaming }
+      .forEach { rpc ->
+        val codeBlock = CodeBlock.of(
+          """
+          return %M(channel, get${rpc.name}Method(), callOptions, request)
+        """.trimIndent(),
```
Comment (oldergod) [inline wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt:40]: indention is weird

File: wire-library/wire-grpc-server-generator/src/test/golden/MethodDescriptor.kt [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package routeguide
+
+import io.grpc.MethodDescriptor
+import kotlin.jvm.Volatile
+
+public class RouteGuideWireGrpc {
+  @Volatile
+  private var getGetFeatureMethod: MethodDescriptor<Point, Feature>? = null
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/golden/MethodDescriptor.kt:8]: consider using Kotlin’s `by lazy` instead of manual double checked locking?

File: wire-library/wire-grpc-server/build.gradle [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    }
+}
+
+dependencies {
+    implementation project(':wire-runtime')
+    // io.grpc.stub relies on guava-android. This module relies on a -jre version of guava.
+    implementation(deps.grpc.stub) {
+        exclude(group: 'com.google.guava', module: 'guava')
+    }
+    implementation ("com.google.guava:guava:21.0")
```
Comment (swankjesse) [inline wire-library/wire-grpc-server/build.gradle:16]: 21 ?

File: wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSinkAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package com.squareup.wire.kotlin.grpcserver
```
Comment (swankjesse) [inline wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSinkAdapter.kt:1]: copyright on these?

File: wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSourceAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
+package com.squareup.wire.kotlin.grpcserver
+
+import com.google.common.util.concurrent.Monitor
+import com.squareup.wire.MessageSource
+import io.grpc.stub.StreamObserver
+
+// This is for adapting Google style grpc stubs to Wire style grpc stubs, in case we want to
```
Comment (swankjesse) [inline wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSourceAdapter.kt:7]: Should we add this to Misk and not Wire?

**Review by swankjesse — APPROVED** [review]
Great PR! Lots of nitpicks. We could land immediately if you like.

**Comment by r3mariano** [comment]
> Great PR! Lots of nitpicks. We could land immediately if you like.

Thanks! I think going through the nitpicks would be good.

File: wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSourceAdapter.kt (hunk 10 above)
Comment (r3mariano) [reply wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSourceAdapter.kt:7]: I think it should be here or a separate library - you don't really need Misk to write up a gRPC service using the generated files (despite the PoC being there), instead it would be a gRPC server and Wire protos.

Comment (r3mariano) [reply wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSourceAdapter.kt:7]: I'll change the code comment though, and try to write a sample to illustrate.

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/ImplBase.kt [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+      FunSpec.builder("bindService")
+        .addModifiers(KModifier.OVERRIDE)
+        .returns(ClassName("io.grpc", "ServerServiceDefinition"))
+        .addCode(bindServiceCodeBlock(service))
+        .build()
+    )
+
+    service.rpcs
+      .flatMap { listOf(it.requestType, it.responseType) }
+      .distinct()
```
[the remark this answers is not in the document]
Comment (r3mariano) [reply wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/ImplBase.kt]: This didn't quite work - reckon flatMap is just chaining the sets into a list.

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+      }
+    return this
+  }
+
+  private fun FunSpec.Builder.addBlockingStubSignature(rpc: Rpc): FunSpec.Builder = this
+    .addParameter("request", ClassName.bestGuess(rpc.requestType.toString()))
+    .returns(
+      if (rpc.requestStreaming)
+        Iterator::class.asClassName()
+          .parameterizedBy(ClassName.bestGuess(rpc.responseType.toString()))
```
[the remark this answers is not in the document]
Comment (r3mariano) [reply wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt]: I've used the type map that KotlinGenerator makes.

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+import com.squareup.wire.schema.Rpc
+import com.squareup.wire.schema.Service
+
+object BlockingStub {
+  internal fun addBlockingStub(builder: TypeSpec.Builder, service: Service): TypeSpec.Builder =
+    builder
+      .addFunction(
+        FunSpec.builder("newBlockingStub")
+          .addParameter("channel", ClassName("io.grpc", "Channel"))
+          .returns(ClassName("", "${service.name}BlockingStub"))
```
[the remark this answers is not in the document]
Comment (r3mariano) [reply wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/BlockingStub.kt]: you're right I forgot to fix this up. i now use the kotlinGenerator type map for this.

File: wire-library/wire-grpc-server/build.gradle (hunk 8 above)
Comment (r3mariano) [reply wire-library/wire-grpc-server/build.gradle:16]: I've kept this at 21 given it's the oldest possible one that supports Monitor, and the rest of Wire is at minimum 20.

File: wire-library/wire-grpc-server-generator/src/test/golden/MethodDescriptor.kt (hunk 7 above)
Comment (r3mariano) [reply wire-library/wire-grpc-server-generator/src/test/golden/MethodDescriptor.kt:8]: This and the get with `@Volatile` are copies of the generated plain Java gRPC code. I'm not sure of the implications of using `by lazy` that might make it different from this (though I think it should be the same).

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/ImplBase.kt [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    }
+  }
+
+  private fun addImplBaseBody(builder: TypeSpec.Builder, service: Service): TypeSpec.Builder {
+    service.rpcs.forEach { rpc ->
+      builder.addFunction(
+        FunSpec.builder(rpc.name)
+          .addModifiers(KModifier.OPEN)
+          .apply { addImplBaseRpcSignature(this, rpc) }
+          .addCode(CodeBlock.of("TODO(\"not implemented\")"))
```
[the remark this answers is not in the document]
Comment (r3mariano) [reply wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/ImplBase.kt]: Yup, done.

**Comment by r3mariano** [comment]
I think I'd like to get this in and iterate on the rest of the nits separately, but I'd also like to write up a sample to verify that the generated code works.

File: wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/KotlinGrpcGenerator.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+import com.squareup.wire.kotlin.grpcserver.BlockingStubGenerator.addBlockingStub
+import com.squareup.wire.kotlin.grpcserver.ImplBaseGenerator.addImplBase
+import com.squareup.wire.kotlin.grpcserver.LegacyAdapterGenerator.addLegacyAdapter
+import com.squareup.wire.kotlin.grpcserver.MethodDescriptorGenerator.addMethodDescriptor
+import com.squareup.wire.kotlin.grpcserver.ServiceDescriptorGenerator.addServiceDescriptor
+import com.squareup.wire.kotlin.grpcserver.StubGenerator.addStub
+import com.squareup.wire.schema.ProtoType
+import com.squareup.wire.schema.Service
+
+class KotlinGrpcGenerator(private val typeToKotlinName: Map<ProtoType, TypeName>) {
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/main/java/com/squareup/wire/kotlin/grpcserver/KotlinGrpcGenerator.kt:30]: thanks for adding that

File: wire-library/wire-grpc-server-generator/src/test/golden/LegacyAdapter.kt [hunk 16]
```diff
@@ -0,0 +1,43 @@
+package routeguide
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/golden/LegacyAdapter.kt:1]: include a note that this is generated code and should not be edited?

File: wire-library/wire-grpc-server-generator/src/test/golden/MethodDescriptor.kt (hunk 7 above)
Comment (swankjesse) [reply wire-library/wire-grpc-server-generator/src/test/golden/MethodDescriptor.kt:8]: When in Kotlin . . .

File: wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/ImplBaseTest.kt [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+              builder = this,
+              service = service
+            )
+          }
+          .build()
+      )
+      .build()
+      .toString()
+
+    println(code)
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/ImplBaseTest.kt:48]: remove?

File: wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/LegacyAdapterTest.kt [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+              builder = this,
+              service
+            )
+          }
+          .build()
+      )
+      .build()
+      .toString()
+
+    println(code)
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/LegacyAdapterTest.kt:49]: remove

File: wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/MethodDescriptorTest.kt [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+            builder = this,
+            service = service,
+            rpc = service.rpc("GetFeature")!!
+          ) }
+          .build()
+      )
+      .build()
+      .toString()
+
+    println(code)
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/MethodDescriptorTest.kt:48]: same

File: wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/ServiceDescriptorTest.kt [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    val code = FileSpec.builder("routeguide", "RouteGuide")
+      .addType(
+        TypeSpec.classBuilder("RouteGuideWireGrpc")
+          .apply { ServiceDescriptorGenerator.addServiceDescriptor(this, service!!) }
+          .build()
+      )
+      .build()
+      .toString()
+
+    println(code)
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/ServiceDescriptorTest.kt:42]: remove

File: wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/StubTest.kt [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+            generator = ClassNameGenerator(buildClassMap(repoBuilder.schema(), service!!)),
+            builder = this,
+            service
+          ) }
+          .build()
+      )
+      .build()
+      .toString()
+
+    println(code)
```
Comment (swankjesse) [inline wire-library/wire-grpc-server-generator/src/test/java/com/squareup/wire/kotlin/grpcserver/StubTest.kt:45]: remove

File: wire-library/wire-grpc-server/build.gradle (hunk 8 above)
Comment (swankjesse) [reply wire-library/wire-grpc-server/build.gradle:16]: We should bump it to something current! But no need for doing that in this PR

File: wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSinkAdapter.kt [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.squareup.wire.MessageSink
+import io.grpc.stub.StreamObserver
+
+// This is for adapting Google style grpc stubs to Wire style grpc stubs.
+class MessageSinkAdapter<T:Any> constructor(private val responseObserver: StreamObserver<T>):
+  MessageSink<T> {
+  override fun cancel() {
+    TODO("not implemented")
```
Comment (swankjesse) [inline wire-library/wire-grpc-server/src/main/java/com/squareup/wire/kotlin/grpcserver/MessageSinkAdapter.kt:25]: is this implementable?

**Review by swankjesse — APPROVED** [review]
